### PR TITLE
Support exporting quantile for summary metrics 

### PR DESCRIPTION
**Pull request by khanhntd**
**Description:** 
As OTEL summary metrics will get [exported as a Statistical Set](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/awsemfexporter/datapoint.go#L174-L189) with `awsemfexporter`, customers will lose the information for each quantile and this would be more critical  if receiver does not set Min and Max with Summary Metric (result in setting Statistical Set will have Min and Max at 0). Therefore, by [turning on](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17265/files#diff-07df5960b9a12ac201861f3d947df425bbbf057f55c8d402a634f39ff99dc6c8R81-R84) `DetailedMetrics`, customers can have detailed information regarding each quantile (the DetailedMetrics config option would allow the same with Histogram Metric since its exported as Statistical Set too)

**Testing:** 
By applying these changes, I have test by:
* Update the[ unit test ](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5eff41478e26af6afc8a45cb8e8fa718c97db001/exporter/awsemfexporter/datapoint_test.go#L342-L417)and running with `go test`
```
ok      github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter       1.020s
```
* Manual testing with the following config
```
exporters:
    awsemf/prometheus:
        dimension_rollup_option: NoDimensionRollup
        eks_fargate_container_insights_enabled: false
        endpoint: ""
        local_mode: false
        log_group_name: Detailed-Prometheus-On
        log_stream_name: '{job}'
        max_retries: 2
        metric_declarations:
            - dimensions:
                - - key1
                  - quantile                  
              label_matchers:
                - label_names:
                    - job
                  regex: MY_JOB
                  separator: ;
              metric_name_selectors:
                - ^go_gc_duration_seconds$
        metric_descriptors: []
        detailed_metrics: true
        namespace: CWAgent-Internal
        no_verify_ssl: false
        num_workers: 8
        output_destination: cloudwatch
        parse_json_encoded_attr_values: []
        proxy_address: ""
        region: ""
        request_timeout_seconds: 30
        resource_arn: ""
        resource_to_telemetry_conversion:
            enabled: true
        role_arn: ""
extensions: {}
processors:
    batch/prometheus:
        send_batch_max_size: 0
        send_batch_size: 8192
        timeout: 200ms
    metricstransform/prometheus:
        transforms: []
    resource/prometheus:
        attributes:
            - action: upsert
              converted_type: ""
              from_attribute: service.name
              from_context: ""
              key: job
              pattern: ""
              value: null
            - action: delete
              converted_type: ""
              from_attribute: ""
              from_context: ""
              key: service.name
              pattern: ""
              value: null
            - action: upsert
              converted_type: ""
              from_attribute: service.instance.id
              from_context: ""
              key: instance
              pattern: ""
              value: null
            - action: delete
              converted_type: ""
              from_attribute: ""
              from_context: ""
              key: service.instance.id
              pattern: ""
              value: null
            - action: delete
              converted_type: ""
              from_attribute: ""
              from_context: ""
              key: net.host.port
              pattern: ""
              value: null
            - action: delete
              converted_type: ""
              from_attribute: ""
              from_context: ""
              key: http.scheme
              pattern: ""
              value: null
            - action: insert
              converted_type: ""
              from_attribute: ""
              from_context: ""
              key: Version
              pattern: ""
              value: 1
            - action: insert
              converted_type: ""
              from_attribute: ""
              from_context: ""
              key: receiver
              pattern: ""
              value: prometheus
receivers:
    prometheus/prometheus:
        buffer_count: 0
        buffer_period: 0s
        config:
            global:
                evaluation_interval: 1m
                scrape_interval: 1m
                scrape_timeout: 10s
            scrape_configs:
                - enable_http2: true
                  file_sd_configs:
                    - files:
                        - /home/ec2-user/prometheus_sd_1.yaml
                      refresh_interval: 5m
                  follow_redirects: true
                  honor_timestamps: true
                  job_name: MY_JOB
                  metrics_path: /metrics
                  sample_limit: 10000
                  scheme: http
                  scrape_interval: 1m
                  scrape_timeout: 10s
        start_time_metric_regex: ""
        target_allocator: null
        use_start_time_metric: false
service:
    extensions: []
    pipelines:
        metrics/prometheus:
            exporters:
                - awsemf/prometheus
            processors:
                - batch/prometheus
                - resource/prometheus
                - metricstransform/prometheus
            receivers:
                - prometheus/prometheus
    telemetry:
        logs:
            development: false
            disable_caller: false
            disable_stacktrace: false
            encoding: json
            error_output_paths: []
            initial_fields: {}
            level: info
            output_paths: []
        metrics:
            address: ""
            level: none
        resource: {}
        traces:
            propagators: []

```

and the result is (generate 6 different logs event) - with quantile changing between 0, 0.25, 0.5, 0.75 and 1 and with count and sum logs event
```
{
    "Version": "1",
    "_aws": {
        "CloudWatchMetrics": [
            {
                "Namespace": "CWAgent-Internal",
                "Dimensions": [
                    [
                        "key1",
                        "quantile"
                    ]
                ],
                "Metrics": [
                    {
                        "Name": "go_gc_duration_seconds"
                    }
                ]
            }
        ],
        "Timestamp": 1672102208350
    },
    "go_gc_duration_seconds": 0.00058318,
    "instance": "127.0.0.1:8000",
    "job": "MY_JOB",
    "key1": "value1",
    "key2": "value2",
    "prom_metric_type": "summary",
    "quantile": "1",
    "receiver": "prometheus"
}
```

```
{
    "Version": "1",
    "go_gc_duration_seconds_count": 1,
    "go_gc_duration_seconds_sum": 0.0000597010000000231,
    "instance": "127.0.0.1:8000",
    "job": "MY_JOB",
    "key1": "value1",
    "key2": "value2",
    "prom_metric_type": "summary",
    "receiver": "prometheus"
}
```

Before apply the changes, the summary metrics will have the following Embedded Metric Format(EMF)
```
 {
    "Version": "1",
    "_aws": {
        "CloudWatchMetrics": [
            {
                "Namespace": "CWAgent-Internal",
                "Dimensions": [
                    [
                        "key1"
                    ]
                ],
                "Metrics": [
                    {
                        "Name": "go_gc_duration_seconds"
                    }
                ]
            }
        ],
        "Timestamp": 1672101428349
    },
    "go_gc_duration_seconds": {
        "Max": 0.00058318,
        "Min": 0.000033009,
        "Count": 0,
        "Sum": 0
    },
    "instance": "127.0.0.1:8000",
    "job": "MY_JOB",
    "key1": "value1",
    "key2": "value2",
    "prom_metric_type": "summary",
    "receiver": "prometheus"
}
```

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: exporter/awsemfexporter/datapoint.go [hunk 1]
```diff
@@ -12,20 +12,32 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/*
+datapoint.go - calculate delta values or keep the value  for each OTEL type (e.g calculate delta for Counter)
+*/
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/datapoint.go:17]:
```suggestion
```

This isn't a package-level godoc string.  Descriptions of functionality like this should go on types or methods.

File: exporter/awsemfexporter/datapoint.go [hunk 2]
```diff
@@ -137,20 +150,19 @@ func (dps numberDataPointSlice) At(i int) (dataPoint, bool) {
...
 	}
 
-	return dataPoint{
-		value:       metricVal,
-		labels:      labels,
-		timestampMs: timestampMs,
-	}, retained
+	datapoints = append(datapoints, dataPoint{name: dps.metricName, value: metricVal, labels: labels, timestampMs: timestampMs})
+
+	return datapoints, retained
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/datapoint.go:155]:
```suggestion
	return []dataPoint{{name: dps.metricName, value: metricVal, labels: labels, timestampMs: timestampMs}}, retained
```

If I'm reading this correctly this is the only time `datapoints` is modified after it is initialized to an empty slice.  Instead of using `append()` here I would just return a slice literal containing the one value.  the return above can return an empty slice literal.

File: exporter/awsemfexporter/datapoint.go [hunk 3]
```diff
@@ -122,13 +134,14 @@ func (dps numberDataPointSlice) At(i int) (dataPoint, bool) {
...
+	var datapoints []dataPoint = []dataPoint{}
+
 	if dps.adjustToDelta {
 		var deltaVal interface{}
-		deltaVal, retained = deltaMetricCalculator.Calculate(dps.metricName, mergeLabels(dps.deltaMetricMetadata, labels),
-			metricVal, metric.Timestamp().AsTime())
+		deltaVal, retained = deltaMetricCalculator.Calculate(dps.metricName, mergeLabels(dps.deltaMetricMetadata, labels), metricVal, metric.Timestamp().AsTime())
 		if !retained {
-			return dataPoint{}, retained
+			return datapoints, retained
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/datapoint.go:144]:
```suggestion
			return nil, retained
```

This doesn't even need to be an allocated slice since `nil` and an empty slice can be used interchangeably in most situations.

File: exporter/awsemfexporter/datapoint.go [hunk 4]
```diff
@@ -122,13 +134,14 @@ func (dps numberDataPointSlice) At(i int) (dataPoint, bool) {
 		metricVal = float64(metric.IntValue())
 	}
 
-	retained := true
+	var retained bool = true
+	var datapoints []dataPoint = []dataPoint{}
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/datapoint.go:138]:
```suggestion
```

Don't think this is needed if the return locations are changed as suggested.

File: exporter/awsemfexporter/datapoint.go [hunk 5]
```diff
@@ -209,44 +232,42 @@ func createLabels(attributes pcommon.Map, instrLibName string) map[string]string
...
+func getDataPoints(pmd pmetric.Metric, metadata cWMetricMetadata, logger *zap.Logger) dataPointSlice {
+	metricMetadata := deltaMetricMetadata{
+		adjustToDelta: false,
+		metricName:    pmd.Name(),
+		namespace:     metadata.namespace,
+		logGroup:      metadata.logGroup,
+		logStream:     metadata.logStream,
 	}
 
+	var dps dataPointSlice = nil
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/datapoint.go:252]:
```suggestion
	var dps dataPointSlice
```

File: exporter/awsemfexporter/emf_exporter_test.go [hunk 6]
```diff
@@ -26,6 +26,8 @@ import (
 	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
 	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
 	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/cwlogs"
+	internaldata "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/opencensus"
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/emf_exporter_test.go:30]: This is not following the import grouping rules used by this repo and will fail the lint checks.  Revert the changes to this file.

File: exporter/awsemfexporter/grouped_metric.go [hunk 7]
```diff
@@ -18,10 +18,9 @@ import (
 	"encoding/json"
 	"strings"
 
+	aws "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/metrics"
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/grouped_metric.go:21]: This is not following the import grouping rules used by this repo and will fail the lint checks.  Revert the import changes to this file.

File: exporter/awsemfexporter/grouped_metric.go [hunk 8]
```diff
@@ -39,68 +38,70 @@ type metricInfo struct {
...
-		}
 
-		metric := &metricInfo{
-			value: dp.value,
-			unit:  translateUnit(pmd, descriptor),
-		}
+			metric := &metricInfo{
+				value: dp.value,
+				unit:  translateUnit(pmd, descriptor),
+			}
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/grouped_metric.go:76]: Are these spaces and not tabs?  Be sure to run `go fmt`.

File: exporter/awsemfexporter/datapoint_test.go [hunk 9]
```diff
@@ -15,575 +15,402 @@
...
+	metrics := rs.ScopeMetrics().AppendEmpty().Metrics()
+	metric := metrics.AppendEmpty()
+	metric.SetName(name)
+	metric.SetUnit("Count")
+	gaugeMetric := metric.SetEmptyGauge()
+	gaugeDatapoint := gaugeMetric.DataPoints().AppendEmpty()
+	gaugeDatapoint.Attributes().PutStr("label1", "value1")
+
+	switch valueType {
+	case "double":
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/datapoint_test.go:46]: Create a distinct type and constants for these.

```go
type gaugeValueType string

const doubleGauge gaugeValueType = "double"
```

and then only use the constants.  This should help avoid typos that accidentally change test behavior.

File: exporter/awsemfexporter/metric_translator.go [hunk 10]
```diff
@@ -20,17 +20,15 @@ import (
 	"reflect"
 	"time"
 
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/cwlogs"
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/metric_translator.go:23]: This is not following the import grouping rules used by this repo and will fail the lint checks.  Revert the import changes to this file.

File: exporter/awsemfexporter/util_test.go [hunk 11]
```diff
@@ -20,13 +20,12 @@ import (
 	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
 	agentmetricspb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/metrics/v1"
 	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	internaldata "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/opencensus"
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/util_test.go:23]: This is not following the import grouping rules used by this repo and will fail the lint checks.  Revert the changes to this file.

File: exporter/awsemfexporter/datapoint.go (hunk 3 above)
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go:141]: I'm fine with that. The reason why I giving its a variable is to keep consistency between calculateDeltaDatapoints function.

File: exporter/awsemfexporter/datapoint.go (hunk 1 above)
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go]: I usually do both but I'm fine with removing it too. Done and thanks Anthony !!!

File: exporter/awsemfexporter/datapoint.go (hunk 5 above)
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go]: Done fixing it !!

File: exporter/awsemfexporter/datapoint.go (hunk 4 above)
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go]: Done !! Thanks for your insight !!!

File: exporter/awsemfexporter/emf_exporter_test.go (hunk 6 above)
Comment (khanhntd) [reply exporter/awsemfexporter/emf_exporter_test.go]: Done too !! Thanks for your insight !!!

File: exporter/awsemfexporter/grouped_metric.go (hunk 7 above)
Comment (khanhntd) [reply exporter/awsemfexporter/grouped_metric.go]: Done too !! Thanks for your insight !!!

File: exporter/awsemfexporter/grouped_metric.go (hunk 8 above)
Comment (khanhntd) [reply exporter/awsemfexporter/grouped_metric.go:77]: Yes I had already run `go fmt`.

File: exporter/awsemfexporter/metric_translator.go (hunk 10 above)
Comment (khanhntd) [reply exporter/awsemfexporter/metric_translator.go]: Done too !! Thanks for your insight !!!

File: exporter/awsemfexporter/metric_translator_test.go [hunk 12]
```diff
@@ -28,15 +28,14 @@ import (
 	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
 	"github.com/golang/protobuf/ptypes/timestamp"
 	"github.com/golang/protobuf/ptypes/wrappers"
+	internaldata "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/opencensus"
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/metric_translator_test.go]: Done too !! Thanks for your insight !!!

File: exporter/awsemfexporter/util_test.go (hunk 11 above)
Comment (khanhntd) [reply exporter/awsemfexporter/util_test.go]: Done too !! Thanks for your insight !!!

File: exporter/awsemfexporter/datapoint.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go]: Done too !! Thanks for your insight !!!

File: exporter/awsemfexporter/datapoint_test.go (hunk 9 above)
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint_test.go]: Done !!! Thanks for your insight Anthony.

File: exporter/awsemfexporter/datapoint.go (hunk 2 above)
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go]: Its the same reason as below
```
I'm fine with that. The reason why I giving its a variable is to keep consistency between calculateDeltaDatapoints function.
```
However, I'm fine with changing this !

File: exporter/awsemfexporter/datapoint.go [hunk 13]
```diff
@@ -109,9 +121,9 @@ type summaryMetricEntry struct {
 }
 
 // At retrieves the NumberDataPoint at the given index and performs rate/delta calculation if necessary.
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go]: Done !! Nice catch on this !!!

File: exporter/awsemfexporter/datapoint.go [hunk 14]
```diff
@@ -43,29 +55,30 @@ func calculateSummaryDelta(prev *aws.MetricValue, val interface{}, timestampMs t
...
 	timestampMs int64
 }
 
-// dataPoints is a wrapper interface for:
+// dataPointSlice is a wrapper interface for:
 //   - pmetric.NumberDataPointSlice
 //   - pmetric.HistogramDataPointSlice
 //   - pmetric.SummaryDataPointSlice
-type dataPoints interface {
+type dataPointSlice interface {
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go:66]: Its to corresponding with the implementation details though ! I'm fine with changing it back too

File: exporter/awsemfexporter/README.md [hunk 15]
```diff
@@ -33,10 +33,12 @@ The following exporter configuration parameters are supported.
...
+| `detailed_metrics`           | "detailed_metrics" is the options for preserving the OTEL metrics values instead of basic metrics (e.g instead of exporting quantile as a Statistical value , the EMF exporter will preserve the quantile's population                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | `false` |
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/README.md:36]: IMO. It's good to me. Will change it and thanks for your comment Evan !

File: exporter/awsemfexporter/datapoint.go [hunk 16]
```diff
@@ -43,29 +55,30 @@ func calculateSummaryDelta(prev *aws.MetricValue, val interface{}, timestampMs t
...
 	Len() int
-	// At gets the adjusted datapoint from the DataPointSlice at i-th index.
+	// CalculateDeltaDatapoints calculates the delta datapoint from the DataPointSlice at i-th index
+	// for some type (Counter, Summary)
 	// dataPoint: the adjusted data point
 	// retained: indicates whether the data point is valid for further process
 	// NOTE: It is an expensive call as it calculates the metric value.
-	At(i int) (dataPoint dataPoint, retained bool)
+	CalculateDeltaDatapoints(i int, instrumentationLibraryName string, detailedMetrics bool) (dataPoint []dataPoint, retained bool)
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go]: Agree. I forgot to notice that we use `ilm.Scope().Name()` instead of the old legacy one. However, nice catch and thanks for noticing though.

File: exporter/awsemfexporter/datapoint.go [hunk 17]
```diff
@@ -85,20 +98,19 @@ func mergeLabels(m deltaMetricMetadata, labels map[string]string) map[string]str
...
 type numberDataPointSlice struct {
-	instrumentationLibraryName string
 	deltaMetricMetadata
 	pmetric.NumberDataPointSlice
 }
 
 // histogramDataPointSlice is a wrapper for pmetric.HistogramDataPointSlice
 type histogramDataPointSlice struct {
-	instrumentationLibraryName string
+	// Todo:(khanhntd) Calculate delta value for count and sum value with histogram
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go:93]: Yup. Good point to me. I have created the issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18245 and added it into the PR. Thanks for catching it out.

File: exporter/awsemfexporter/metric_translator.go [hunk 18]
```diff
@@ -28,9 +28,8 @@ import (
 
 const (
 	// OTel instrumentation lib name as dimension
-	oTellibDimensionKey          = "OTelLib"
-	defaultNamespace             = "default"
-	noInstrumentationLibraryName = "Undefined"
+	oTellibDimensionKey = "OTelLib"
+	defaultNamespace    = "default"
```
Comment (Aneurysm9) [inline exporter/awsemfexporter/metric_translator.go:32]: It's not clear to me why the changes in this file are made.  How do they relate to exporting quantiles?

Comment (khanhntd) [reply exporter/awsemfexporter/metric_translator.go:32]: I delete the `noInstrumentationLibraryName= Undefined ` since [it is being used to know](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17265/files#diff-7ab9fa22126292a4cb18c74ec6309729f3fc05049b7e5889dd80266db71f8f43R222-R223) if there are any instrumental library and the `fmt` doing the rest. However, we can check if the value is empty instead of passing the value `Undefinied`

**Review by Aneurysm9 — APPROVED** [review]

File: exporter/awsemfexporter/datapoint.go [hunk 19]
```diff
@@ -43,22 +53,24 @@ func calculateSummaryDelta(prev *aws.MetricValue, val interface{}, timestampMs t
...
 // dataPoint represents a processed metric data point
 type dataPoint struct {
+	name        string
 	value       interface{}
 	labels      map[string]string
 	timestampMs int64
 }
 
-// dataPoints is a wrapper interface for:
+// dataPointSlice is a wrapper interface for:
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/datapoint.go]: I think its an old commit?

File: exporter/awsemfexporter/README.md [hunk 20]
```diff
@@ -33,10 +33,12 @@ The following exporter configuration parameters are supported.
...
+| `detailed_metrics`           | "detailed_metrics" is the options for retaining detailed datapoint values in exported metrics (e.g instead of exporting a quantile as a statistical value, preserve the quantile's population)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | `false` |
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply exporter/awsemfexporter/README.md]: I'm fine with removing it since I saw others document such as output_destination having the same pattern
```
| `output_destination`                         | "output_destination" is an option to specify the EMFExporter output. Currently, two options are available. "cloudwatch" or "stdout"  
```

File: exporter/awsemfexporter/datapoint.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (evan-bradley) [reply exporter/awsemfexporter/datapoint.go]: Whoops, I missed that. Looks good, then.

**Review by evan-bradley — APPROVED** [review]

**Review by codeboten — APPROVED** [review]
